MapboxCoreNavigation/RouteController.swift

```diff
@@ -390,10 +390,23 @@ extension RouteController: CLLocationManagerDelegate {
         
         // Bearings need to normalized so when the `finalHeading` is 359 and the user heading is 1,
         // we count this as within the `RouteControllerMaximumAllowedDegreeOffsetForTurnCompletion`
-        if let finalHeading = routeProgress.currentLegProgress.upComingStep?.finalHeading {
+        if let upcomingStep = routeProgress.currentLegProgress.upComingStep, let finalHeading = upcomingStep.finalHeading, let initialHeading = upcomingStep.initialHeading {
+            let initialHeadingNormalized = wrap(initialHeading, min: 0, max: 360)
             let finalHeadingNormalized = wrap(finalHeading, min: 0, max: 360)
             let userHeadingNormalized = wrap(location.course, min: 0, max: 360)
-            courseMatchesManeuverFinalHeading = differenceBetweenAngles(finalHeadingNormalized, userHeadingNormalized) <= RouteControllerMaximumAllowedDegreeOffsetForTurnCompletion
+            let differenceBetweenInitialAndFinalHeading = differenceBetweenAngles(initialHeadingNormalized, finalHeadingNormalized)
+            
+            // If the upcoming maneuver is fairly straight,
+            // do not check if the user is within x degrees of the exit heading.
+            // For ramps, their current heading will very close to the exit heading.
+            // We need to wait until their moving away from the maneuver location instead.
+            // We can do this by looking at their snapped distance from the maneuver.
+            // Once this distance is zero, they are at more moving away from the maneuver location
+            if differenceBetweenInitialAndFinalHeading <= RouteControllerMaximumAllowedDegreeOffsetForTurnCompletion {
+                courseMatchesManeuverFinalHeading = userSnapToStepDistanceFromManeuver == 0
```

Comment:
Intersections in OpenStreetMap are somewhat idealized. What if a sharp left is modeled like this:

<img width="291" alt="sharp-left" src="https://user-images.githubusercontent.com/1231218/28147337-9e1fdba0-6734-11e7-993d-5a074ad5601c.png">

The user’s location never snaps to the precisely the end of the road. Do we have a provision for preceding to the next step if the user is closer to the next step geometry than the current step geometry?